_posts/2020-07-21-cube-functionality-to-kartothek.markdown

```diff
@@ -0,0 +1,414 @@
+---
+layout: single
+title:  "Introducing Cube Functionality To Kartothek"
+date:   2020-07-21 06:00:00 +0500
+tags: technology python data-engineering
+header:
+  overlay_image: assets/images/tech_gear_banner.jpg
+  overlay_filter: 0.2
+  show_overlay_excerpt: false
+author: Marco Neumann & Usha Nemani
+author_profile: true
+---
+# Introducing Cube Functionality To Kartothek
+
+In our last blog we introduced [Kartothek](2019-05-28-introducing-kartothek.markdown), a table management python library powered by Dask. 
+Our journey continued by adding a gem to our jewel. We empowered Kartothek with multiple-dataset functionality. 
+Kartothek already provides Dataset features, As a user do I really need a multiple-dataset? Hang On!! 
+Let us spend rest of our time understanding the story of **Cube (kartothek that supports multiple datasets)**. 
+
+## What is a Cube?
+A Cube deals with multiple Kartothek datasets.
+
+Let us start with building a cube for **geodata**. Similar to Kartothek, 
+we need a [simplekv](https://simplekv.readthedocs.io/)-based store backend along with an abstract cube definition.
+
+**df_weather** is a pandas dataframe created from reading a csv file.
+
+```python
+>>> from io import StringIO
+>>> df_weather = pd.read_csv(
+...     filepath_or_buffer=StringIO("""
+... avg_temp     city country        day
+...        6  Hamburg      DE 2018-01-01
+...        5  Hamburg      DE 2018-01-02
+...        8  Dresden      DE 2018-01-01
+...        4  Dresden      DE 2018-01-02
+...        6   London      UK 2018-01-01
+...        8   London      UK 2018-01-02
+...     """.strip()),
+...     delim_whitespace=True,
+...     parse_dates=["day"],
+... )
+```
+
+
+```python
+>>> from kartothek.core.cube.cube import Cube
+>>>##we are creating a geodata cube instance
+>>> cube = Cube(
+...     uuid_prefix="geodata",
+...     dimension_columns=["city", "day"],
+...     partition_columns=["country"]
+...)
+```
+
+We use the simple **kartothek.io.eager_cube** backend to store the data:
+
+```python
+>>> from kartothek.io.eager_cube import build_cube
+>>> datasets_build = build_cube(
+...   data=df_weather,
+...   store=store,
+...   cube=cube
+...)
+```
+
+
+where **store** is the **simplekv** store of storefactory. (For more details, please refer our [Kartothek](2019-05-28-introducing-kartothek.markdown) blog.)
+
+We just have preserved a single Kartothek dataset. Lets print the content of seed dataset.
+
+```python
+>>> print(", ".join(sorted(datasets_build.keys())))
+seed
+>>> ds_seed = datasets_build["seed"].load_all_indices(store)
+>>> print(ds_seed.uuid)
+geodata++seed
+>>> print(", ".join(sorted(ds_seed.indices)))
+city, country, day
+```
+
+Finally, let’s have a quick look at the store content. Note that we cut out UUIDs and timestamps here for documentation purposes:
```

Comment:
```suggestion
Now we have a quick look at the store content. Note that we cut out UUIDs and timestamps here for documentation purposes:
```

It's not "finally" since it is not the last section of the post.

Reply:
Updated.